src/Compilers/CSharp/Test/Semantic/Semantics/InterceptorsTests.cs

```diff
@@ -0,0 +1,871 @@
+            {
+                [Interceptable]
+                public I1 InterceptableMethod(string param) { Console.Write("interceptable " + param); return this; }
+            }
+
+            static class Program
+            {
+                public static void Main()
+                {
+                    var c = new C();
+                    c.InterceptableMethod("call site");
+                }
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 15, 10)]
+                public static I1 Interceptor1(this I1 i1, string param) { Console.Write("interceptor " + param); return i1; }
+            }
+            """;
+        var comp = CreateCompilation(new[] { (source, "Program.cs"), s_attributesSource });
+        comp.VerifyEmitDiagnostics(
+            // Program.cs(22,6): error CS27011: Interceptor must have a 'this' parameter matching parameter 'C this' on 'C.InterceptableMethod(string)'.
+            //     [InterceptsLocation("Program.cs", 15, 10)]
+            Diagnostic(ErrorCode.ERR_InterceptorMustHaveMatchingThisParameter, @"InterceptsLocation(""Program.cs"", 15, 10)").WithArguments("C this", "C.InterceptableMethod(string)").WithLocation(22, 6)
+            );
+    }
+
+    // PROTOTYPE(ic): test interceptable explicit interface implementation (should error).
+    // PROTOTYPE(ic): test interceptor with 'ref this' to match a struct interceptable method.
+
+    [Fact]
+    public void SignatureMismatch_03()
+    {
+        // Instance method 'this' parameter ref kind differs from interceptor 'this' parameter ref kind.
+        var source = """
+            using System.Runtime.CompilerServices;
+            using System;
+
+            struct S
+            {
+                [Interceptable]
+                public void InterceptableMethod(string param) { Console.Write("interceptable " + param); }
+            }
+
+            static class Program
+            {
+                public static void Main()
+                {
+                    var s = new S();
+                    s.InterceptableMethod("call site");
+                }
+            }
+
+            static class D
+            {
+                [InterceptsLocation("Program.cs", 14, 10)]
+                public static void Interceptor1(this S s, string param) { Console.Write("interceptor " + param); }
+            }
+            """;
+        var comp = CreateCompilation(new[] { (source, "Program.cs"), s_attributesSource });
+        comp.VerifyEmitDiagnostics(
+            // Program.cs(21,6): error CS27011: Interceptor must have a 'this' parameter matching parameter 'ref S this' on 'S.InterceptableMethod(string)'.
+            //     [InterceptsLocation("Program.cs", 14, 10)]
+            Diagnostic(ErrorCode.ERR_InterceptorMustHaveMatchingThisParameter, @"InterceptsLocation(""Program.cs"", 14, 10)").WithArguments("ref S this", "S.InterceptableMethod(string)").WithLocation(21, 6)
+            );
+    }
```

Comment:
>} [](http://example.com/codeflow?start=4&length=1)

Possible test cases:
- Intercept instance method with instance method in same class, base class, derived class
- Intercept with extern method
- Intercept with abstract or interface method
- Intercept an abstract or interface method
- Intercept a virtual or overridden method
- Intercept a non-extension call to a static method with a static method when one or both are extension methods
- An explicit interface implementation marked as interceptable
- Intercept a generic method call when the type parameters are / are not substituted

Reply:
Adding these in a PROTOTYPE comment. Thank you for these, they all seem very useful.